### ISSUE TITLE
Accelerate Newton assembly

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#2274]
@MFraters sent me profiles for Newton-Stokes computations that look like this (in release mode):
```
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |  1.16e+05s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |       351 |       425s |      0.37% |
| Assemble Stokes system newton   |      7304 |  6.47e+04s |        56% |
| Assemble Stokes system picard   |       748 |  1.55e+03s |       1.3% |
| Assemble Stokes system rhs      |      7701 |  3.15e+03s |       2.7% |
| Assemble composition system     |      8052 |  4.13e+03s |       3.6% |
| Assemble temperature system     |      8052 |  5.47e+03s |       4.7% |
| Build Stokes preconditioner     |      8052 |  2.11e+04s |        18% |
| Build composition preconditioner|      8052 |       559s |      0.48% |
| Build temperature preconditioner|      8052 |       474s |      0.41% |
| Solve Stokes system             |      8052 |  1.22e+04s |        11% |
| Solve composition system        |      8052 |       324s |      0.28% |
| Solve temperature system        |      8052 |       306s |      0.26% |
| Free surface                    |       352 |       117s |       0.1% |
| Initialization                  |         1 |      1.19s |         0% |
| Postprocessing                  |       351 |      77.5s |         0% |
| Refine mesh structure, part 1   |       352 |       139s |      0.12% |
| Refine mesh structure, part 2   |       352 |      48.9s |         0% |
| Setup dof systems               |       353 |       225s |      0.19% |
| Setup initial conditions        |         2 |      2.08s |         0% |
+---------------------------------+-----------+------------+------------+
```
Note how the Newton assembly requires 56% of the run time. We should look into why this is so.

For comparison, on the same test case when using the Picard iteration, he gets this:
```
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |  1.66e+05s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |     34142 |   4.2e+04s |        25% |
| Assemble composition system     |     34142 |  1.77e+04s |        11% |
| Assemble temperature system     |     34142 |  2.34e+04s |        14% |
| Build Stokes preconditioner     |     34142 |  4.54e+04s |        27% |
| Build composition preconditioner|     34142 |  2.41e+03s |       1.5% |
| Build temperature preconditioner|     34142 |  2.02e+03s |       1.2% |
| Solve Stokes system             |     34142 |  2.96e+04s |        18% |
| Solve composition system        |     34142 |  1.18e+03s |      0.71% |
| Solve temperature system        |     34142 |   1.2e+03s |      0.72% |
| Free surface                    |       352 |       124s |         0% |
| Initialization                  |         1 |      1.25s |         0% |
| Postprocessing                  |       351 |      98.7s |         0% |
| Refine mesh structure, part 1   |       352 |       140s |         0% |
| Refine mesh structure, part 2   |       352 |      49.2s |         0% |
| Setup dof systems               |       353 |       227s |      0.14% |
| Setup initial conditions        |         2 |      2.03s |         0% |
+---------------------------------+-----------+------------+------------+
```

**Comment by bangerth** [comment]
Also of interest: The time to assemble one Stokes system with Picard is 1.2 seconds. For one Newton step, it is almost 9 seconds.

**Comment by MFraters** [comment]
closed by #2381, #2465, #2470 and #2496.
